### ISSUE TITLE
Enhancement: [no-unnecessary-type-arguments] Detect inferred type

### ISSUE THREAD

**Issue by Samuel-Beslogic** [typescript-eslint/typescript-eslint#4978]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-type-arguments

### Description

Assuming this is feasible, I would like redundant type arguments due to inferred type to be flagged as well. Even if the type argument is required.

The examples below uses a `MatDialog` from `@angular/material/dialog`. Both result in `dialogRef: MatDialogRef<RemoveItemDialogComponent, any>`

```js
export interface ComponentType<T> {
    new (...args: any[]): T;
}
export declare class MatDialogRef<T, R = void> {
   // ...
}
// ...
  open<T, D = any, R = any>(component: ComponentType<T>, config?: MatDialogConfig<D>): MatDialogRef<T, R>;
```

### Fail

```typescript
// Complex case: Inferred type will be the same
const dialogRef = this.matDialog.open<RemoveItemDialogComponent>(RemoveItemDialogComponent);

// Simple case: Is already a string
const a_string = '' as string;
const fn = <T>(arg: T) => arg;
const result = fn<string>(a_string);
```

### Pass

```typescript
// Complex case
const dialogRef = this.matDialog.open(RemoveItemDialogComponent);

// Simpler cases
const fn = <T>(arg: T) => arg;
const a_string = '' as string;

// result is string
const result = fn(a_string);

// Changes '' to string
const result = fn<string>('');
// or (opinionated)
const result: string = fn('');
```

**Comment by JoshuaKGoldberg** [comment]
Agreed, I would love this rule. See also older discussion in #295, particularly https://github.com/typescript-eslint/typescript-eslint/issues/295#issuecomment-688526380.

That issue was closed because inferable type annotations can help TypeScript performance -- but much like the [`explicit-module-boundary-types`](https://typescript-eslint.io/rules/explicit-module-boundary-types) rule, those only come into play in larger and/or more complex codebases. I'd be strongly in favor of us making a rule for it with a caveat on performance.

Unfortunately this is yet another rule blocked on TypeScript exposing an assignability API: https://github.com/microsoft/TypeScript/issues/9879.

**Comment by JoshuaKGoldberg** [comment]
Now that v8 is released, per https://github.com/typescript-eslint/typescript-eslint/discussions/7936, we can mark this issue as unblocked! 🚀 

I'm wondering, would this be a big enough increase in the rule's power to be considered a breaking change? cc @typescript-eslint/triage-team - I would consider this really 49% a _bug_ in that the rule _should_ have caught these things already. So, I'm happy considering treating this as non-breaking change.

**Comment by ej-shafran** [comment]
Has anyone moved this issue forward? If not - I can clear up some type on my schedule in the next couple weeks and try to implement this. I would really like this functionality in my projects.

**Comment by JoshuaKGoldberg** [comment]
Nope, no progress. It'd likely be linked here if it did exist. The floor is yours!

We don't use any kind of issue claiming system as we've found they result in "licked cookie" situations: https://typescript-eslint.io/contributing/issues/#issue-claiming

**Comment by ej-shafran** [comment]
I can try and give this another go, unless y'all think I should leave it alone 😅 (I know there's no claiming issues, but I just wanna make sure I haven't disappointed too bad before I give it another go).

My thoughts on this are:
- Maybe the issues can be avoided by declaring this as a separate rule? This doesn't actually resolve our problem, but it seems like people are pretty divided on what they want this rule to do...
- For this to work well (at least, like I imagined it) we need something more like `checker.isTypeIdenticalTo` (which isn't actually exported by the TypeScript compiler API) to validate if the types are actually the same or just look similar. That, or we need some way to create a "phantom" version of the call/type instantiation without the explicit type arguments, and grab what TypeScript _would_ infer about this statement from that...?

Apologies for the mess! Hopefully the final implementation can include some tests relating to the issues people have pointed out...

**Comment by bradzacher** [comment]
I think the biggest issue with the previous implementation was that it used assignability to check instead of exact types.

It's perfectly reasonable to want to use a named type instead of an anonymous one, for example, but the rule was reporting on that.

Put another way:
- it is acceptable to _narrow_ the type using a type argument. For example using a named type instead of an anonymous object type.
- it is acceptable to _widen_ the type using a type argument. For example widening a string literal type to the `string` type.
- it is **unacceptable** to use the exact same type.

**Comment by JoshuaKGoldberg** [comment]
> can try and give this another go

💪 Yes! Please do! This is an exciting area of investigation that we don't have the bandwidth on the team to take on ourselves. And even if we did, it's great having community folks work on it. Better to share the learnings that come from working on things than to keep them isolated with us.

> disappointed ... the mess

_The mess is good_! This rule is a really, really tricky one. Plus now it's exponentially tricker with the addition of assignability APIs, which are still very new to usage in linting (or anybody outside of TypeScript, really). Trying things out like this and seeing what happens is an important and natural part of the process.

<details>
<summary>An unnecessary shpiel on responsibility in open source</summary>

The final responsibility for preventing bugs is on open source maintainers, not external contributors. We're stewards of the project and are the ones with the most context on it. External contributors can't and shouldn't be trusted to fully vet their changes: that's way too much ramp-up and context discovery to expect of them!

Our ideal would be that multiple team members review each PR, _especially_ tricky ones like #11918. Unfortunately, we -like many projects in open source- are grossly understaffed. If we blocked every PR on that then we'd almost never accept anything, far fewer contributors & contributions would happen, and the project would stagnate.

As the sole reviewer of your PR, that means it was _my_ slip-up that caused the bugs. Not yours. And I probably could & should have caught some of the bugs -- especially thinking to check about the type parameter counts. But, again, that's on me. Not you.
</details>

tl;dr: you're doing good work and we'd very much like to see it continue 🚀 

> avoided by declaring this as a separate rule? 

We don't generally have two somewhat-overlapping rules made available. The common practice would be to have an option. I'd suggest one here that changes whether the rule uses the old, basic, not-very-many-cases-caught logic vs. the new assignability one. That way we can give folks the lenient behavior as a default for now, allow enterprising individuals to opt into the stricter behavior if they want, and plan to change the default to the stricter behavior in the next major version. And perhaps even enable it in the `strict` config but not `recommended`.

> something more like `checker.isTypeIdenticalTo`
> grab what TypeScript _would_ infer about this statement from that

😬 As nice as either or both would be, the APIs available to us from TypeScript are pretty much frozen. You might be able to get some of the way there by checking for referential equality (`typeA === typeB`).

I'll jot some notes on strategies later. 

> can include some tests relating to the issues people have pointed out...

Agreed. For quick reference, here are the issues I could find, with some quick notes:

* #12157: _(I haven't investigated the crash)_
* #12158: Blagh, nested `any`s. These are always a pain. Perhaps this could detect that the type parameter (`TArgs`) has a default type (`Args`) that includes an `any`? Rules often give up on analysis when an `any` is involved, so this wouldn't be unusual.
  * #12046 is a reference of another place where nested `any`s are looked for.
* #12170: tl;dr, and not an isolated reproduction. I'd ignore this one.
* #12183: Blagh again, another nested `any`. Perhaps this could detect that a function parameter (`sideEffect`) has a reference to a type parameter (the `T` in `newValue: T`), and detect that the parameter's provided argument (`anyFn`) has an `any` in it? 
* #12195: `foo`'s declaration/symbol could be read to see that third type parameter is required.
* #12196: For function overloads, there might be some heuristic that could be used to determine whether a type argument is necessary. Perhaps if overloads differ in their type arguments?

Several of those are duplicates but they still show somewhat different situations in code. IMO they should all have explicit tests, just to be safe.
